File: 785-is-graph-bipartite/785-is-graph-bipartite.java
class Solution {
    //-1 : not visited , 0-red color ,1-- green color
    public boolean isBipartite(int[][]graph,int src,int[]vis){
       LinkedList<Integer>q=new LinkedList<>();
        
       int color=0;
        
       q.addLast(src);
        
        while(q.size()>0){
            int size=q.size();
            while(size-->0){
                int vtx=q.removeFirst();
                
                if(vis[vtx]!=-1){
                    if(vis[vtx]!=color)return false;
                }
                
                vis[vtx]=color;
                
                for(int v:graph[vtx]){
                    if(vis[v]==-1){
                        q.addLast(v);
                    }
                }
                
            }
            color=(color+1)%2;
        }
        
        return true;
        
    }
    
    public boolean isBipartite(int[][] graph) {
        int n=graph.length;
        
        int[]vis=new int[n];
        Arrays.fill(vis,-1);
        
        for(int i=0;i<n;i++){
            if(vis[i]==-1 && !isBipartite(graph,i,vis))return false;
        }
        return true;
    }
}

File: 785-is-graph-bipartite/785-is-graph-bipartite.cpp
class Solution {
public:
    
    bool bfs_bipartite(vector<vector<int>>& graph,vector<int>&colors,int src){
        int c=0; // lets start with the blue color
        
        queue<int>q;
        q.push(src);
        
        while(q.size()){
            int size=q.size();
            
            while(size--){
                int u=q.front();q.pop();
                
                if(colors[u]!=-1){
                    if(colors[u]!=c)return false;
                    continue;
                }
                
                colors[u]=c;
                
                for(int v:graph[u]){
                    if(colors[v]==-1){
                        q.push(v);
                    }
                }
            }
            c=(c+1)%2;
        }
        
        return true;
    }
    
    bool isBipartite(vector<vector<int>>& graph) {
        int V=graph.size();
        
        vector<int>colors(V,-1);
        
        for(int i=0;i<V;i++){
            if(colors[i]==-1){
                bool isBipartite=bfs_bipartite(graph,colors,i);
                
                if(!isBipartite)return false;
            }
        }
        return true;
    }
};


// -1 --> unvisited
// 0 --> blue
//1 --> green